### PR TITLE
Yago/step5

### PR DESCRIPTION
**Pull request by YagoAccenture**
Step 5:
When clicking on a character, we will navigate to a screen with detailed information about that character (Name, status, image, origin and location) as well as a button to return to the previous screen

## Summary by CodeRabbit

- **New Features**
  - Enabled view binding for improved UI management.
  - Added `DetailsActivity` to display character details.

- **Enhancements**
  - Updated `Character` data class to include detailed location and origin information.
  - Modified `CharacterAdapter` to handle item click events efficiently.
  - Added new string resources for "Origin" and "Location."

- **UI Improvements**
  - Updated `details_screen.xml` layout with new views and adjusted constraints for better display of character details.

- **Dependencies**
  - Added `androidx.recyclerview` for enhanced list handling in the app.

### CODE REVIEW

File: app/src/main/java/com/mstudio/superheroarch/CharacterAdapter.kt [hunk 1]
```diff
@@ -10,25 +10,42 @@ import com.squareup.picasso.Picasso
...
+    class CharacterViewHolder(view: View, listener: onItemClickListener) : RecyclerView.ViewHolder(view) {
         val name: TextView = view.findViewById(R.id.character_name)
         val status: TextView = view.findViewById(R.id.character_status)
         val image: ImageView = view.findViewById(R.id.character_image)
+
+        init {
+            itemView.setOnClickListener{
+                listener.onItemClick(adapterPosition)
+            }
+        }
```
[the remark this answers is not in the document]
Comment (PabloGrelaM) [reply app/src/main/java/com/mstudio/superheroarch/CharacterAdapter.kt:32]: adapterPosition is deprecated. The bot is right here @YagoAccenture

File: app/src/main/java/com/mstudio/superheroarch/MainActivity.kt [hunk 2]
```diff
@@ -20,12 +21,22 @@ class MainActivity : AppCompatActivity() {
     private val adapter = CharacterAdapter()
     private lateinit var apiRick: ApiRick
 
+    companion object {
+        const val EXTRA_NAME = "com.mstudio.superheroarch.MainActivity.EXTRA_NAME"
+        const val EXTRA_STATUS = "com.mstudio.superheroarch.MainActivity.EXTRA_STATUS"
+        const val EXTRA_IMAGE = "com.mstudio.superheroarch.MainActivity.EXTRA_IMAGE"
+        const val EXTRA_LOCATION = "com.mstudio.superheroarch.MainActivity.EXTRA_LOCATION"
+        const val EXTRA_ORIGIN= "com.mstudio.superheroarch.MainActivity.EXTRA_ORIGIN"
+
```
[the remark this answers is not in the document]
Comment (PabloGrelaM) [reply app/src/main/java/com/mstudio/superheroarch/MainActivity.kt:30]: The bot is right, but use Serializable.

File: app/build.gradle.kts [hunk 3]
```diff
@@ -7,6 +7,10 @@ android {
     namespace = "com.mstudio.superheroarch"
     compileSdk = 34
 
+    viewBinding{
```
Comment (PabloGrelaM) [inline app/build.gradle.kts:10]: you are not using this.

File: app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -19,6 +19,9 @@
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
+        <activity android:name=".DetailsActivity"
+            android:exported="true">
```
Comment (PabloGrelaM) [inline app/src/main/AndroidManifest.xml:23]: why exported?

File: app/src/main/java/com/mstudio/superheroarch/Character.kt [hunk 5]
```diff
@@ -1,9 +1,24 @@
...
 
+import android.database.Observable
+import java.util.Objects
+
 data class Character(
     val name: String,
     val status: String,
-    val image: String
+    val image: String,
+    val origin: CharacterOrigin,
```
Comment (PabloGrelaM) [inline app/src/main/java/com/mstudio/superheroarch/Character.kt:10]: What's the difference between CharacterOrigin and CharacterLocation?

File: app/src/main/res/layout/details_screen.xml [hunk 6]
```diff
@@ -4,39 +4,185 @@
...
+        app:layout_constraintStart_toStartOf="parent"
+        android:backgroundTint="#00000000"
+        android:scaleType="centerCrop"
+        android:src="@drawable/backimg"
+        />
+
+
+
+
+   <!--
```
Comment (PabloGrelaM) [inline app/src/main/res/layout/details_screen.xml:25]: it's forbidden to upload commented code.

File: app/src/main/res/layout/details_screen.xml [hunk 7]
```diff
@@ -4,39 +4,185 @@
...
-        tools:text="StatusTest"
+        tools:text="Alive"
         app:layout_constraintTop_toBottomOf="@id/character_name"
         app:layout_constraintStart_toStartOf="parent"
         app:layout_constraintEnd_toEndOf="parent"/>
 
 
+    <androidx.cardview.widget.CardView
+        android:id="@+id/cardlocation"
+        android:layout_width="match_parent"
```
Comment (PabloGrelaM) [inline app/src/main/res/layout/details_screen.xml:80]: Why do you have 0dp in the image and match_parent here?

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -7,4 +7,7 @@
     <string name="no_character_found">Failed to find character</string>
     <string name="name">Name: </string>
     <string name="status">Status: </string>
+    <string name="origin">Origin</string>
+    <string name="location">Location</string>
+    <string name="back">\u2190</string>
```
Comment (PabloGrelaM) [inline app/src/main/res/values/strings.xml:12]: \u2190???

**Review by PabloGrelaM — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/mstudio/superheroarch/CharacterAdapter.kt [hunk 9]
```diff
@@ -10,25 +10,42 @@ import com.squareup.picasso.Picasso
...
 class CharacterAdapter : RecyclerView.Adapter<CharacterAdapter.CharacterViewHolder>() {
 
-    private var characters: MutableList<Character> = mutableListOf()
+    private lateinit var mListener: onItemClickListener
+    interface onItemClickListener{
+        fun onItemClick(position: Int)
+    }
+    fun setOnItemClickListener(listener: onItemClickListener){
+        mListener = listener
+    }
```
[the remark this answers is not in the document]
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/CharacterAdapter.kt:19]: Solved.

File: app/src/main/AndroidManifest.xml (hunk 4 above)
Comment (YagoAccenture) [reply app/src/main/AndroidManifest.xml:23]: I put it on to see what the screen would look like without running the application flow, I forgot to remove it.
Immediately removed

File: app/build.gradle.kts (hunk 3 above)
Comment (YagoAccenture) [reply app/build.gradle.kts:10]: True, it was the first way I used to do step 5, but I did not eliminate it when I stopped using it.

File: app/src/main/AndroidManifest.xml [hunk 10]
```diff
@@ -19,6 +19,9 @@
...
+        </activity>
```
[the remark this answers is not in the document]
Comment (YagoAccenture) [reply app/src/main/AndroidManifest.xml:24]: Solved.

File: app/src/main/java/com/mstudio/superheroarch/Character.kt (hunk 5 above)
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/Character.kt:10]: They are 2 different objects in the API response, they contain the same, a url and a location, but the values of the origin and the location can vary

File: app/src/main/java/com/mstudio/superheroarch/MainActivity.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/MainActivity.kt:30]: Fixed in the next commit

File: app/src/main/java/com/mstudio/superheroarch/CharacterAdapter.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/CharacterAdapter.kt:32]: Fixed in the next commit

**Review by PabloGrelaM — CHANGES_REQUESTED** [review]
Still with some things wrong.

File: app/build.gradle.kts [hunk 11]
```diff
@@ -7,6 +7,8 @@ android {
     namespace = "com.mstudio.superheroarch"
     compileSdk = 34
 
+
```
Comment (PabloGrelaM) [inline app/build.gradle.kts:10]: Don't add empty lines to the code.

**Review by PabloGrelaM — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/details_screen.xml (hunk 7 above)
Comment (YagoAccenture) [reply app/src/main/res/layout/details_screen.xml]: Fixed, now it takes up 80% of the screen and is no longer match parent.

File: app/src/main/res/layout/details_screen.xml [hunk 12]
```diff
@@ -4,39 +4,169 @@
...
         android:layout_height="wrap_content"
-        android:padding="16dp"
-        tools:text="NombreTest"
+        android:paddingTop="4dp"
+        tools:text="Rick Sanchez"
         app:layout_constraintEnd_toEndOf="parent"
         app:layout_constraintStart_toStartOf="parent"
         app:layout_constraintTop_toBottomOf="@+id/character_image"
+        android:textStyle="bold"
+        android:textSize="20dp"
```
[the remark this answers is not in the document]
Comment (YagoAccenture) [reply app/src/main/res/layout/details_screen.xml:45]: Solved in next commit.

File: app/src/main/java/com/mstudio/superheroarch/Character.kt [hunk 13]
```diff
@@ -1,10 +1,26 @@
 package com.mstudio.superheroarch
 
+import android.database.Observable
```
[the remark this answers is not in the document]
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/Character.kt]: Solved in next commit

File: app/build.gradle.kts (hunk 11 above)
Comment (YagoAccenture) [reply app/build.gradle.kts:10]: Solved in next commit

File: app/src/main/java/com/mstudio/superheroarch/Character.kt [hunk 14]
```diff
@@ -1,10 +1,19 @@
 package com.mstudio.superheroarch
 
+import java.io.Serializable
```
[the remark this answers is not in the document]
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/Character.kt:3]: Yes it is in use

File: app/src/main/java/com/mstudio/superheroarch/Character.kt [hunk 15]
```diff
@@ -1,10 +1,26 @@
...
+
 data class Character(
     val name: String,
     val status: String,
-    val image: String
-)
+    val image: String,
+    val origin: CharacterOrigin,
+    val location: CharacterLocation
+) : Serializable
```
[the remark this answers is not in the document]
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/Character.kt]: Solved in next commit

File: app/src/main/java/com/mstudio/superheroarch/CharacterAdapter.kt [hunk 16]
```diff
@@ -10,25 +10,41 @@ import com.squareup.picasso.Picasso
...
+            }
+        }
     }
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CharacterViewHolder {
         val view = LayoutInflater.from(parent.context).inflate(R.layout.item_view, parent, false)
-        return CharacterViewHolder(view)
+        return mListener?.let {
+            CharacterViewHolder(view, it)
+        } ?: throw IllegalStateException("Listener no puede ser nulo. Asegúrate de llamar a setOnItemClickListener antes de crear el adaptador.")
```
[the remark this answers is not in the document]
Comment (YagoAccenture) [reply app/src/main/java/com/mstudio/superheroarch/CharacterAdapter.kt]: Solved in next commit

**Review by PabloGrelaM — APPROVED** [review]

**Pull request merged**